supernova/src/ccs/lccs.rs

```diff
@@ -0,0 +1,448 @@
+use ark_crypto_primitives::sponge::CryptographicSponge;
+use ark_ec::{AdditiveGroup, CurveGroup};
+use ark_ff::Field;
+use ark_poly::{
+    DenseMVPolynomial, DenseMultilinearExtension, MultilinearExtension, SparseMultilinearExtension,
+};
+use ark_poly_commit::{LabeledPolynomial, PolynomialCommitment};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+
+use ark_std::ops::Index;
+use std::ops::Neg;
+
+#[cfg(feature = "parallel")]
+use rayon::iter::{IntoParallelIterator, IntoParallelRefIterator, ParallelIterator};
+
+use super::super::r1cs::R1CSShape;
+pub use super::super::sparse::{MatrixRef, SparseMatrix};
+use super::mle::{fold_vec_to_mle_low, matrix_to_mle, mle_to_mvp, vec_to_mle};
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error {
+    InvalidWitnessLength,
+    InvalidInputLength,
+    InvalidEvaluationPoint,
+    InvalidTargets,
+    FailedWitnessCommitting,
+    NotSatisfied,
+}
+
+pub struct LCCSShape<G: CurveGroup> {
+    /// `m` in the CCS/HyperNova papers.
+    pub num_constraints: usize,
+    /// Witness length.
+    ///
+    /// `m - l - 1` in the CCS/HyperNova papers.
+    pub num_vars: usize,
+    /// Length of the public input `X`. It is expected to have a leading
+    /// `ScalarField` element (`u`), thus this field must be non-zero.
+    ///
+    /// `l + 1`, w.r.t. the CCS/HyperNova papers.
+    pub num_io: usize,
+    /// Number of matrices.
+    ///
+    /// `t` in the CCS/HyperNova papers.
+    pub num_matrices: usize,
+    /// Number of multisets.
+    ///
+    /// `q` in the CCS/HyperNova papers.
+    pub num_multisets: usize,
+    /// Max cardinality of the multisets.
+    ///
+    /// `d` in the CCS/HyperNova papers.
+    pub max_cardinality: usize,
+    /// Set of constraint matrices.
+    pub Ms: Vec<SparseMultilinearExtension<G::ScalarField>>,
```

Comment:
I think making this change will help reduce the frustration with different endianness for polynomial types. We will only need to evaluate the polynomials `MLE(MZ)` in hypernova, rather than `MLE(M)` or `MLE(Z)`. Having the "column variables" be vectors rather than polynomials keeps them unambiguous.